File: viewpagerdotsindicator/src/main/java/com/tbuonomo/viewpagerdotsindicator/OnPageChangeListenerHelper.java
package com.tbuonomo.viewpagerdotsindicator;

import android.support.v4.view.ViewPager;

public abstract class OnPageChangeListenerHelper implements ViewPager.OnPageChangeListener {
  private int currentPage;
  private int lastPage;

  @Override public final void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
    int selectedPosition = currentPage;
    if (position != currentPage && positionOffset == 0 || currentPage < position) {
      resetPosition(selectedPosition);
      selectedPosition = currentPage = position;
    }

    if (Math.abs(currentPage - position) > 1) {
      resetPosition(selectedPosition);
      currentPage = lastPage;
    }

    int nextPosition = -1;
    if (currentPage == position && currentPage + 1 < getPageCount()) {
      nextPosition = currentPage + 1;
    } else if (currentPage > position) {
      nextPosition = selectedPosition;
      selectedPosition = currentPage - 1;
    }

    onPageScrolled(selectedPosition, nextPosition, positionOffset);

    lastPage = position;
  }

  @Override public final void onPageSelected(int position) {
    currentPage = position;
  }

  @Override public final void onPageScrollStateChanged(int i) {

  }

  abstract void onPageScrolled(int selectedPosition, int nextPosition, float positionOffset);

  abstract void resetPosition(int position);

  abstract int getPageCount();
}

File: viewpagerdotsindicator/src/main/java/com/tbuonomo/viewpagerdotsindicator/OnPageChangeListenerHelper.kt
package com.tbuonomo.viewpagerdotsindicator

abstract class OnPageChangeListenerHelper {
    private var lastLeftPosition: Int = -1
    private var lastRightPosition: Int = -1

    internal abstract val pageCount: Int

    fun onPageScrolled(position: Int, positionOffset: Float) {
        var offset = (position + positionOffset)
        val lastPageIndex = (pageCount - 1).toFloat()
        if (offset == lastPageIndex) {
            offset = lastPageIndex - .0001f
        }
        val leftPosition = offset.toInt()
        val rightPosition = leftPosition + 1

        if (rightPosition > lastPageIndex || leftPosition == -1) {
            return
        }

        onPageScrolled(leftPosition, rightPosition, offset % 1)

        if (lastLeftPosition != -1) {
            if (leftPosition > lastLeftPosition) {
                (lastLeftPosition until leftPosition).forEach {
                    resetPosition(it)
                }
            }

            if (rightPosition < lastRightPosition) {
                resetPosition(lastRightPosition)
                ((rightPosition + 1)..lastRightPosition).forEach {
                    resetPosition(it)
                }
            }
        }

        lastLeftPosition = leftPosition
        lastRightPosition = rightPosition
    }

    internal abstract fun onPageScrolled(
        selectedPosition: Int, nextPosition: Int,
        positionOffset: Float
    )

    internal abstract fun resetPosition(position: Int)
}
